build-tools/packages/build-cli/src/codeCoverage/codeCoveragePr.ts

```diff
@@ -0,0 +1,92 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { getAzureDevopsApi } from "@fluidframework/bundle-size-tools";
+import { type IAzureDevopsBuildCoverageConstants } from "../library/azureDevops/constants.js";
+import {
+	getBaselineBuildMetrics,
+	getBuildArtifactForSpecificBuild,
+} from "../library/azureDevops/getBaselineBuildMetrics.js";
+import type { CommandLogger } from "../logging.js";
+import { compareCodeCoverage } from "./compareCodeCoverage.js";
+import { getCommentForCodeCoverageDiff } from "./getCommentForCodeCoverage.js";
+import { getCoverageMetricsForBaseline } from "./getCoverageMetrics.js";
+
+/**
+ * Summary of code coverage analysis.
+ */
+export interface CodeCoverageSummary {
+	/**
+	 * Message to be put in the comment.
+	 */
+	commentMessage: string;
+
+	/**
+	 * Whether to fail the build or not.
+	 */
+	failBuild: boolean;
+}
+
+/**
+ * API to get the code coverage summary for a PR.
+ * @param adoToken - ADO token that will be used to download artifacts from ADO pipeline runs.
+ * @param codeCoverageConstantsBaseline - The code coverage constants required for fetching the baseline build artifacts.
+ * @param codeCoverageConstantsPR - The code coverage constants required for fetching the PR build artifacts.
+ * @param changedFiles - The list of files changed in the PR.
+ * @param logger - The logger to log messages.
+ */
+export const getCodeCoverageSummary = async (
+	adoToken: string,
+	codeCoverageConstantsBaseline: IAzureDevopsBuildCoverageConstants,
+	codeCoverageConstantsPR: IAzureDevopsBuildCoverageConstants,
+	changedFiles: string[],
+	logger?: CommandLogger,
+): Promise<CodeCoverageSummary> => {
+	const adoConnection = getAzureDevopsApi(adoToken, codeCoverageConstantsBaseline.orgUrl);
+
+	const baselineBuildInfo = await getBaselineBuildMetrics(
+		codeCoverageConstantsBaseline,
+		adoConnection,
+		logger,
+	);
+
+	if (baselineBuildInfo === undefined || typeof baselineBuildInfo === "string") {
+		return {
+			commentMessage: baselineBuildInfo ?? "No Baseline build found",
+			failBuild: false,
+		};
+	}
+
+	const adoConnectionForPR = getAzureDevopsApi(adoToken, codeCoverageConstantsPR.orgUrl);
+
+	const pRBuildInfo = await getBuildArtifactForSpecificBuild(
+		codeCoverageConstantsPR,
+		adoConnectionForPR,
+		logger,
+	);
+
+	if (pRBuildInfo === undefined || typeof pRBuildInfo === "string") {
+		return {
+			commentMessage: pRBuildInfo ?? "No PR build found",
+			failBuild: false,
+		};
+	}
+
+	// Extract the coverage metrics for the baseline and PR builds.
+	const [coverageMetricsForBaseline, coverageMetricsForPr] = await Promise.all([
+		getCoverageMetricsForBaseline(baselineBuildInfo.artifactZip),
+		getCoverageMetricsForBaseline(pRBuildInfo.artifactZip),
```

Comment:
The artifact that it uses gets published from the cobertura tool results which runs before we run this step of comparison. So, we do already have that artifact published.

Reply:
Can you point to a successful run of the whole workflow that confirms that? I could be wrong, but I feel pretty confident that's not the case. The `PublishCodeCoverageResults@2` task in line 398 of `build-npm-package.yml` publishes something for the ADO UI, but not the artifact that we can download by name; that one gets published on line 579, after pretty much everything we do in the pipeline.

Reply:
Yes, I think you are right because I was running the flow locally because I cannot run in CI until I release and consume new build tools packages and since the PR on which I am running the flow locally has once already published that artifact, I am able to run it.

But that would be another PR, so I will open another PR for this.